File: Python/String/M-1190. Reverse Substrings Between Each Pair of Parentheses.py
class Solution:
    @classmethod
    def reverseParentheses(self, s: str) -> str:
        stack = [[]]
        for c in s:
            if c == "(":
                stack.append([])
                continue

            if c == ")":
                cur = stack.pop()
                cur.reverse()
                stack[len(stack) - 1].extend(cur)
                continue

            stack[len(stack) - 1].append(c)

        return "".join(stack[0])


testData = [
    {
        "s": "(abcd)",
        "ans": "dcba",
    },
    {
        "s": "(u(love)i)",
        "ans": "iloveu",
    },
    {
        "s": "(ed(et(oc))el)",
        "ans": "leetcode",
    },
]

for data in testData:
    print(data["ans"] == Solution.reverseParentheses(data["s"]))

File: Javascript/String/M-1190. Reverse Substrings Between Each Pair of Parentheses.ts
function reverseParentheses(s: string): string {
  const stack: string[][] = [[]]; // index 0 supposed to be the result
  for (const c of s) {
    if (c === '(') {
      stack.push([]);
      continue;
    }
    if (c === ')') {
      const curSubstr = stack.pop()?.reverse();
      if (curSubstr) stack[stack.length - 1].push(...curSubstr);
      continue;
    }
    stack[stack.length - 1].push(c);
  }

  return stack[0].join('');
}

const testData = [
  {
    s: '(abcd)',
    ans: 'dcba',
  },
  {
    s: '(u(love)i)',
    ans: 'iloveu',
  },
  {
    s: '(ed(et(oc))el)', // (ed(etco)el) => (edocteel) => leetcode
    ans: 'leetcode',
  },
];

for (const data of testData) {
  console.log(data.ans === reverseParentheses(data.s));
}
